### ISSUE TITLE
Identical Softmax Probabilities Within Minibatches

### ISSUE THREAD

**Issue by RishabGargeya** [dmlc/mxnet#416]
Hi! I'd like to report a puzzling bug that I'm getting in my CNN in which the softmax outputs of each minibatch are all identical. Here's what I mean:
Let's say I have a batch size of 32 for a binary classification task, so the output of my network is a probability array of size (32, 2). Between minibatch iterations, the probabilities are updated, yet each pair of probabilities within the minibatch corresponding to the 32 images in the batch is identical.

My code for the neural net architecture is below:

```
def ConvFactory(data, num_filter, kernel=(3,3), stride=(1,1), pad=(1, 1), act_type='leaky'):
    conv = mx.symbol.Convolution(data=data, num_filter=num_filter, kernel=kernel, stride=stride, pad=pad)
    act = mx.symbol.LeakyReLU(data = conv, slope = 0.5)
    return act

def PoolFactory(data):
    pool = mx.symbol.Pooling(data=data, kernel=(3, 3), stride=(2, 2), pool_type='max')
    return pool

data = mx.symbol.Variable(name="data")
conv1 = ConvFactory(data=data, num_filter=32, stride=(2,2), pad=(0,0))
pool1 = PoolFactory(conv1)
conv2 = ConvFactory(pool1, num_filter=32)
conv3 = ConvFactory(conv2, num_filter=32)
pool2 = PoolFactory(conv3)
conv4 = ConvFactory(pool2, num_filter=64)
conv5 = ConvFactory(conv4, num_filter=64)
pool3 = PoolFactory(conv5)
conv6 = ConvFactory(pool3, num_filter=128)
conv7 = ConvFactory(conv6, num_filter=128)
pool4 = PoolFactory(conv7)
conv8 = ConvFactory(pool4, num_filter=256)
conv9 = ConvFactory(conv8, num_filter=256)
conv10 = ConvFactory(conv9, num_filter=256)
pool5 = PoolFactory(conv10)
flatten = mx.symbol.Flatten(data=pool5, name="flatten1")
fc1 = mx.symbol.FullyConnected(data=flatten, num_hidden=1024, name="fc1")

act1 = mx.symbol.LeakyReLU(data=fc1, slope=0.5, 'leaky')

fc2 = mx.symbol.FullyConnected(data=act1, num_hidden=1024, name="fc2")

act2 = mx.symbol.LeakyReLU(data=fc2, slope=0.5, 'leaky')

fc3 = mx.symbol.FullyConnected(data=fc2, num_hidden=2, name="fc3")
softmax = mx.symbol.Softmax(data=fc3, name="loss")

```

As well, here is a sample softmax output for a minibatch iteration, printed from a custom eval function. The first array is the ground truth labels. The (32,2) shaped array is the softmax probabilities.

```
[0 0 1 0 0 1 0 0 1 0 0 1 0 1 0 0 0 1 0 0 0 0 0 0 1 1 1 0 0 0 0 0]
[[ 0.6135819   0.38641813]
 [ 0.6135819   0.38641813]
 [ 0.6135819   0.38641813]
 [ 0.6135819   0.38641813]
 [ 0.6135819   0.38641813]
 [ 0.6135819   0.38641813]
 [ 0.6135819   0.38641813]
 [ 0.6135819   0.38641813]
 [ 0.6135819   0.38641813]
 [ 0.6135819   0.38641813]
 [ 0.6135819   0.38641813]
 [ 0.6135819   0.38641813]
 [ 0.6135819   0.38641813]
 [ 0.6135819   0.38641813]
 [ 0.6135819   0.38641813]
 [ 0.6135819   0.38641813]
 [ 0.6135819   0.38641813]
 [ 0.6135819   0.38641813]
 [ 0.6135819   0.38641813]
 [ 0.6135819   0.38641813]
 [ 0.6135819   0.38641813]
 [ 0.6135819   0.38641813]
 [ 0.6135819   0.38641813]
 [ 0.6135819   0.38641813]
 [ 0.6135819   0.38641813]
 [ 0.6135819   0.38641813]
 [ 0.6135819   0.38641813]
 [ 0.6135819   0.38641813]
 [ 0.6135819   0.38641813]
 [ 0.6135819   0.38641813]
 [ 0.6135819   0.38641813]
 [ 0.6135819   0.38641813]]

```

Thanks for your help!

**Comment by piiswrong** [comment]
Thanks for the bug report.
Could you check if your input to softmax (e.g. fc3) has different values for each row?
you may want to use group output and asnumpy() for printing

**Comment by tqchen** [comment]
@RishabGargeya It would be great if you can use an example dataset (either randomly generated numpy array, or mnist) that reproduce the problem, so we can help you better by looking into what is going on. Thanks.

**Comment by RishabGargeya** [comment]
Hi, Very sorry for the late response. I've tried to reproduce the problem using mnist with a similar (but downsized) architecture for your convenience. However, for some reason the code does not run when using the leaky relu activation function (as opposed to any other activation function). I get the following error:

```
[01:46:21] ./dmlc-core/include/dmlc/logging.h:208: [01:46:21] ./mshadow/mshadow/./tensor_blob.h:198: Check failed: (dim) == (ndim_) dimension do not match target dimension 4 vs 2
[01:46:21] ./dmlc-core/include/dmlc/logging.h:208: [01:46:21] src/engine/./threaded_engine.h:295: [01:46:21] ./mshadow/mshadow/./tensor_blob.h:198: Check failed: (dim) == (ndim_) dimension do not match target dimension 4 vs 2
terminate called after throwing an instance of 'dmlc::Error'
  what():  [01:46:21] src/engine/./threaded_engine.h:295: [01:46:21] ./mshadow/mshadow/./tensor_blob.h:198: Check failed: (dim) == (ndim_) dimension do not match target dimension 4 vs 2
Aborted
```
I was able to replicate this error by simply changing line 22 in this lenet mnist script: (https://github.com/dmlc/mxnet/blob/39d5f364b287a695775a32be46f22e7dfaad583c/example/mnist/lenet.py#L22), to the following:
```
tanh3 = mx.symbol.LeakyReLU(data=fc1, act_type="leaky")
```
Is this an issue with the flatten function?
Thanks you for the responses!

**Comment by tqchen** [comment]
@antinucleon  can you look into this?

**Comment by antinucleon** [comment]
I made a fix: https://github.com/dmlc/mxnet/pull/452
